lib/client/known_hosts_migrate.go

```diff
@@ -0,0 +1,159 @@
+// formatted as an old-style entry. Only old-style entries are candidates for
+// pruning; all others are passed through untouched.
+func isOldStyleHostsEntry(entry *knownHostEntry) bool {
+	if entry.marker != "" {
+		return false
+	}
+
+	if len(entry.hosts) != 1 {
+		return false
+	}
+
+	if entry.comment != "" {
+		return false
+	}
+
+	return true
+}
+
+// canPruneOldHostsEntry determines if a particular old-style hosts entry has an
+// equivalent new-style entry and can thus be pruned.
+func canPruneOldHostsEntry(oldEntry *knownHostEntry, newEntries []*knownHostEntry) bool {
+	// Note: Per sshd's documentation, it is valid (though not recommended) for
+	// repeated/overlapping entries to exist for a given host; as such, it's
+	// only safe to prune an old entry when both the (*.)hostname and public key
+	// match.
+
+	// Note: we already know this entry matches exactly 1 host. The new-style
+	// entries prepend `*.`, so we'll add that upfront.
+	oldHost := "*." + oldEntry.hosts[0]
+
+	// We'll need to marshal the keys so we can compare them properly.
+	oldKey := oldEntry.pubKey.Marshal()
+
+	for _, newEntry := range newEntries {
+		if oldEntry.pubKey.Type() != newEntry.pubKey.Type() {
+			continue
+		}
+
+		newKey := newEntry.pubKey.Marshal()
+		if !bytes.Equal(oldKey, newKey) {
+			continue
+		}
+
+		for _, newHost := range newEntry.hosts {
+			if newHost == oldHost {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// pruneOldHostKeys removes all old-style host keys for which a new-style
+// duplicate entry exists. This may modify order of host keys, but will not
+// change their content.
+func pruneOldHostKeys(output []string) []string {
+	log := logrus.WithField(trace.Component, teleport.ComponentMigrate)
+
+	var (
+		oldEntries   = make([]*knownHostEntry, 0)
+		newEntries   = make([]*knownHostEntry, 0)
+		prunedOutput = make([]string, 0)
+	)
+
+	// First, categorize all existing known hosts entries.
+	for i, line := range output {
+		parsed, err := parseKnownHost(line)
+		if err != nil {
+			// If the line isn't parseable, pass it through.
+			log.Debugf("Unable to parse known host on line %d, skipping", i+1)
+			prunedOutput = append(prunedOutput, line)
+			continue
+		}
+
+		if isOldStyleHostsEntry(parsed) {
+			// Only old-style entries are candidates for removal.
+			oldEntries = append(oldEntries, parsed)
+		} else {
+			// Everything else is passed through as-is...
+			prunedOutput = append(prunedOutput, line)
+
+			// ...but only new-style entries are candidates for comparison.
+			if parsed.marker == "cert-authority" {
+				newEntries = append(newEntries, parsed)
+			}
+		}
+	}
+
+	// Next, for each old-style entry, determine if an existing new-style entry
+	// exists. If not, pass it through.
+	for _, entry := range oldEntries {
+		if canPruneOldHostsEntry(entry, newEntries) {
+			log.Debugf("Pruning old known_hosts entry for %s.", entry.hosts[0])
```

Comment:
Check hosts length?

Reply:
Old entries have their length checked in `isOldStyleHostsEntry()` and are guaranteed to have exactly 1 host entry. Additionally,  `crypto/ssh`'s `ParseKnownHosts` rejects lines without any hosts so the list should never be empty to begin with.

If this is still problematic for code style reasons I'm happy to add more checks, though.